kopykat-ksp/src/main/kotlin/at/kopyk/KopyKatProcessor.kt

```diff
@@ -1,129 +1,70 @@
-import org.apache.commons.io.FilenameUtils.wildcardMatch
+
+internal interface LoggerScope {
+  val logger: KSPLogger
+}
+
+internal interface OptionsScope {
+  val options: KopyKatOptions
+}
+
+internal class ProcessorScope(
+  environment: SymbolProcessorEnvironment,
+) : LoggerScope, OptionsScope {
+  val codegen = environment.codeGenerator
+  override val logger = environment.logger
+  override val options = KopyKatOptions(environment.logger, environment.options)
+}
 
 internal class KopyKatProcessor(
-  private val codegen: CodeGenerator,
-  private val logger: KSPLogger,
-  private val options: KopyKatOptions
+  private val scope: ProcessorScope,
 ) : SymbolProcessor {
 
-  @OptIn(KspExperimental::class)
   override fun process(resolver: Resolver): List<KSAnnotated> {
-    resolver.getAllFiles().let { files ->
-      if (files.none { it.hasGeneratedMarker() }) {
-        val declarations = files.flatMap { it.allNestedDeclarations() }
-
-        val classes = declarations
-          .filterIsInstance<KSClassDeclaration>()
-          .onEach { it.checkKnown() }
-          .onEach { it.checkRedundantAnnotation() }
-          .filter { it.shouldGenerate() && it.typeCategory is Known }
-
-        // add different copies to data and value classes
-        classes
-          .let { targets -> targets.map { ClassCompileScope(it, classes, logger) } }
-          .forEachRun { process() }
-        // add different copies to type aliases
-        declarations
-          .filterIsInstance<KSTypeAlias>()
-          .onEach { it.checkKnown() }
-          .filter { it.isAnnotationPresent(CopyExtensions::class) && it.typeCategory is Known }
-          .let { targets -> targets.map { TypeAliasCompileScope(it, classes, logger) } }
-          .forEachRun { process() }
-        // add copy from parent to all classes
-        declarations
-          .filterIsInstance<KSClassDeclaration>()
-          .filter { !it.isAbstract() && it.isConstructable() }
-          .forEach {
-            with(ClassCompileScope(it, classes, logger)) {
-              if (options.superCopy) copyFromParentKt.writeTo(codegen)
-            }
+    scope.processFiles(resolver) {
+      // add copy functions for data, value classes, and type aliases
+      if (options.copyMap || options.mutableCopy) {
+        (classes.mapRun { classScope } + typeAliases.mapRun { typealiasScope })
+          .filter { it.canHaveCopyFunctions(options.hierarchyCopy) }
+          .onEachRun { logger.logging("Processing ${simpleName.asString()}") }
+          .forEachRun {
+            if (options.copyMap) copyMapFunctionKt.write()
+            if (options.mutableCopy) mutableCopyKt.write()
           }
       }
+      // add isomorphic copies
+      declarations
+        .filterIsInstance<KSClassDeclaration> { hasCopyAnnotation() }
+        .flatMapRun { classScope.allCopies }
+        .distinctBy { it.name }
+        .let { others -> others.mapNotNull(::fileSpec.partially1(others)) }
```

Comment:
I am really not fond of `partially1` and friends, I find them quite hard to read. But if you prefer it that way, also fine by me.

Reply:
Is this also the only place where we use Arrow? I don't mind having it as a dependency, but having only to partially apply a function seems like a big step.

Reply:
We also have a couple of `zip` for nullables.